## Notebook 5: Seatwork

<b>Instructions:</b> Submit your answers by <b>LT</b>.

<u>LT</u>: <b>10</b> 

<u>Names</u>:\
<b>
&emsp;Janica Lae Batang\
&emsp;Mary Honeylie Buitre\
&emsp;Paolo Jose de Guzman\
&emsp;John Francis Gacal\
&emsp;Vincent John Rivera</b>

For this seatwork, we will be applying the methods we learned to solve systems of equations to constrained optimization problems.

Use the method of Lagrange multipliers to solve the following constrained optimization problem.

\begin{equation}
    \min_\mathbf{x} f(\mathbf{x}) = 81x_1^2 + x_2^2
\end{equation}

subject to

\begin{align}
    4x_1^2 + x_2^2 &= 9
\end{align}

#### Q1.

Setup the Langrangian function and the system of first order conditions.

#### Answer.

\begin{align}
    \mathcal{L}(\mathbf{x}, \lambda) = 81x_1^2 + x_2^2 - \lambda(4x_1^2 + x_2^2 - 9)
\end{align}

\begin{align}
    \frac{\partial \mathcal{L}}{\partial x_1} = 162x_1 -8\lambda x_1 = 0\\
    \frac{\partial \mathcal{L}}{\partial x_2} = 2x_2 - 2\lambda x_2 = 0\\
    \frac{\partial \mathcal{L}}{\partial \lambda} = 4x_1^2 + x_2^2 - 9 = 0\\
\end{align}

#### Q2.

Find the optimums of $f(x)$ by solving the above system of nonlinear equations using the Newton-Raphson method.

<i>Hint</i>: There are 2 maximums and 2 minimums. You'll have to experiment with different initial guesses to find them.

In [3]:
# Insert code here
def f(x):
    return np.array([162*x[0] - 8*x[0]*x[2],
                     2*x[1] - 2*x[1]*x[2],
                     4*x[0]**2 + x[1]**2 - 9])

def f_jacob(x):
    return np.array([[162 - 8*x[2], 0, -8*x[0]],
                     [0, 2-2*x[2], -2*x[1]],
                     [8*x[0], 2*x[1], 0]])

In [4]:
x_k = np.array([1,1,1])
for k in range (0,10):
    print("k = %d --" %k,"-- x_k[0] = %.6f --" %x_k[0],"-- x_k[1] = %.6f --" %x_k[1],"-- x_k[2] = %.6f --" %x_k[2],"-- f(x_k) =", f(x_k))
    x_k = x_k - np.matmul(np.linalg.inv(f_jacob(x_k)), f(x_k))

k = 0 -- -- x_k[0] = 1.000000 -- -- x_k[1] = 1.000000 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [154   0  -4]
k = 1 -- -- x_k[0] = 0.000000 -- -- x_k[1] = 7.000000 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [ 0.  0. 40.]
k = 2 -- -- x_k[0] = 0.000000 -- -- x_k[1] = 4.142857 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.         0.         8.16326531]
k = 3 -- -- x_k[0] = 0.000000 -- -- x_k[1] = 3.157635 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.         0.         0.97066175]
k = 4 -- -- x_k[0] = 0.000000 -- -- x_k[1] = 3.003935 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.         0.         0.02362391]
k = 5 -- -- x_k[0] = 0.000000 -- -- x_k[1] = 3.000003 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.00000000e+00 0.00000000e+00 1.54618955e-05]
k = 6 -- -- x_k[0] = 0.000000 -- -- x_k[1] = 3.000000 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.00000000e+00 0.00000000e+00 6.64357458e-12]
k = 7 -- -- x_k[0] = 0.000000 -- -- x_k[1] = 3.000000 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0. 0. 0.]
k = 8 -- -- x_k[0]

In [5]:
x_k = np.array([-1,-1,1])
for k in range (0,10):
    print("k = %d --" %k,"-- x_k[0] = %.6f --" %x_k[0],"-- x_k[1] = %.6f --" %x_k[1],"-- x_k[2] = %.6f --" %x_k[2],"-- f(x_k) =", f(x_k))
    x_k = x_k - np.matmul(np.linalg.inv(f_jacob(x_k)), f(x_k))

k = 0 -- -- x_k[0] = -1.000000 -- -- x_k[1] = -1.000000 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [-154    0   -4]
k = 1 -- -- x_k[0] = 0.000000 -- -- x_k[1] = -7.000000 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [ 0.  0. 40.]
k = 2 -- -- x_k[0] = 0.000000 -- -- x_k[1] = -4.142857 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.         0.         8.16326531]
k = 3 -- -- x_k[0] = 0.000000 -- -- x_k[1] = -3.157635 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.         0.         0.97066175]
k = 4 -- -- x_k[0] = 0.000000 -- -- x_k[1] = -3.003935 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.         0.         0.02362391]
k = 5 -- -- x_k[0] = 0.000000 -- -- x_k[1] = -3.000003 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.00000000e+00 0.00000000e+00 1.54618955e-05]
k = 6 -- -- x_k[0] = 0.000000 -- -- x_k[1] = -3.000000 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0.00000000e+00 0.00000000e+00 6.64357458e-12]
k = 7 -- -- x_k[0] = 0.000000 -- -- x_k[1] = -3.000000 -- -- x_k[2] = 1.000000 -- -- f(x_k) = [0. 0. 0.]
k = 8 

In [6]:
x_k = np.array([5,5,5])
for k in range (0,15):
    print("k = %d --" %k,"-- x_k[0] = %.6f --" %x_k[0],"-- x_k[1] = %.6f --" %x_k[1],"-- x_k[2] = %.6f --" %x_k[2],"-- f(x_k) =", f(x_k))
    x_k = x_k - np.matmul(np.linalg.inv(f_jacob(x_k)), f(x_k))

k = 0 -- -- x_k[0] = 5.000000 -- -- x_k[1] = 5.000000 -- -- x_k[2] = 5.000000 -- -- f(x_k) = [610 -40 116]
k = 1 -- -- x_k[0] = 71.466667 -- -- x_k[1] = -272.466667 -- -- x_k[2] = 222.973333 -- -- f(x_k) = [-115903.68711111  120960.66844444   94659.02222222]
k = 2 -- -- x_k[0] = 38.344418 -- -- x_k[1] = -133.510351 -- -- x_k[2] = 114.205028 -- -- f(x_k) = [-28821.2069024   30228.08599085  23697.19124099]
k = 3 -- -- x_k[0] = 21.991283 -- -- x_k[1] = -63.550308 -- -- x_k[2] = 60.319960 -- -- f(x_k) = [-7049.51848058  7539.60339254  5964.10769299]
k = 4 -- -- x_k[0] = 14.107623 -- -- x_k[1] = -27.538418 -- -- x_k[2] = 34.614689 -- -- f(x_k) = [-1621.21290283  1851.39073408  1545.46456473]
k = 5 -- -- x_k[0] = 9.845235 -- -- x_k[1] = -8.212557 -- -- x_k[2] = 24.590056 -- -- f(x_k) = [-341.83096737  387.46934399  446.16071641]
k = 6 -- -- x_k[0] = 5.710455 -- -- x_k[1] = -0.876374 -- -- x_k[2] = 22.072727 -- -- f(x_k) = [-83.26880963  36.9351684  122.20521373]
k = 7 -- -- x_k[0] = 3.067484

In [7]:
x_k = np.array([-5,-5,5])
for k in range (0,15):
    print("k = %d --" %k,"-- x_k[0] = %.6f --" %x_k[0],"-- x_k[1] = %.6f --" %x_k[1],"-- x_k[2] = %.6f --" %x_k[2],"-- f(x_k) =", f(x_k))
    x_k = x_k - np.matmul(np.linalg.inv(f_jacob(x_k)), f(x_k))

k = 0 -- -- x_k[0] = -5.000000 -- -- x_k[1] = -5.000000 -- -- x_k[2] = 5.000000 -- -- f(x_k) = [-610   40  116]
k = 1 -- -- x_k[0] = -71.466667 -- -- x_k[1] = 272.466667 -- -- x_k[2] = 222.973333 -- -- f(x_k) = [ 115903.68711111 -120960.66844444   94659.02222222]
k = 2 -- -- x_k[0] = -38.344418 -- -- x_k[1] = 133.510351 -- -- x_k[2] = 114.205028 -- -- f(x_k) = [ 28821.2069024  -30228.08599085  23697.19124099]
k = 3 -- -- x_k[0] = -21.991283 -- -- x_k[1] = 63.550308 -- -- x_k[2] = 60.319960 -- -- f(x_k) = [ 7049.51848058 -7539.60339254  5964.10769299]
k = 4 -- -- x_k[0] = -14.107623 -- -- x_k[1] = 27.538418 -- -- x_k[2] = 34.614689 -- -- f(x_k) = [ 1621.21290283 -1851.39073408  1545.46456473]
k = 5 -- -- x_k[0] = -9.845235 -- -- x_k[1] = 8.212557 -- -- x_k[2] = 24.590056 -- -- f(x_k) = [ 341.83096737 -387.46934399  446.16071641]
k = 6 -- -- x_k[0] = -5.710455 -- -- x_k[1] = 0.876374 -- -- x_k[2] = 22.072727 -- -- f(x_k) = [ 83.26880963 -36.9351684  122.20521373]
k = 7 -- -- x_k[0] = -3.

In [11]:
answers = [[0,3],
           [0,-3],
           [-1.5,0],
           [1.5,0]]

for i in answers:
    print(f'x1 = {i[0]}, x2 = {i[1]}')
    val = 81*i[0]**2 + i[1]**2
    print(f'f(X) = {val}\n')

x1 = 0, x2 = 3
f(X) = 9

x1 = 0, x2 = -3
f(X) = 9

x1 = -1.5, x2 = 0
f(X) = 182.25

x1 = 1.5, x2 = 0
f(X) = 182.25



#### Answer.

\begin{align}
    \mathbf{x}_{\text{max}}^* = (-1.5, 0) \quad &, \quad \mathbf{x}_{\text{max}}^* = (1.5, 0)\\
    \mathbf{x}_{\text{min}}^* = (0, -3) \quad &, \quad \mathbf{x}_{\text{min}}^* = (0, 3)
\end{align}

\begin{align}
    f(\mathbf{x}_{\text{max}}^*) &= 182.25 \\
    f(\mathbf{x}_{\text{min}}^*) &= 9 \\
\end{align}

#### Check.

Solve the constrained optimization problem above using the `scipy.optimize` package.

In [12]:
from scipy.optimize import minimize

def obj(x, sign=1):
    return sign*(81*x[0]**2 + x[1]**2)

def cons(x):
    return 4*x[0]**2 + x[1]**2 - 9

cons = ({'type': 'eq',
         'fun' : cons})

g = [[-1,-1],[1,1]]
g1 = [[-5,-5],[5,5]]

print('Minimum')
for i in g:
    print(f'Initial guess = {i}')
    check = minimize(obj, i, constraints=cons)
    print(f'Function value: {round(check.fun,2)}')
    print(f'x: {np.round(check.x,2)}\n')
    
print('\nMaximum')
for i in g1:
    print(f'Initial guess = {i}')
    check = minimize(obj, i, args=(-1,), constraints=cons)
    print(f'Function value: {round(-check.fun,2)}')
    print(f'x: {np.round(check.x,2)}\n')

Minimum
Initial guess = [-1, -1]
Function value: 9.0
x: [-0. -3.]

Initial guess = [1, 1]
Function value: 9.0
x: [-0.  3.]


Maximum
Initial guess = [-5, -5]
Function value: 182.25
x: [-1.5  0. ]

Initial guess = [5, 5]
Function value: 182.25
x: [ 1.5 -0. ]

